pkg/chains/storage/ca/ca.go

```diff
@@ -0,0 +1,350 @@
+/*
+Copyright 2020 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ca
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	attestationpb "github.com/grafeas/grafeas/proto/v1beta1/attestation_go_proto"
+	commonpb "github.com/grafeas/grafeas/proto/v1beta1/common_go_proto"
+	pb "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	"github.com/pkg/errors"
+	"github.com/tektoncd/chains/pkg/artifacts"
+	"github.com/tektoncd/chains/pkg/chains/formats"
+	"github.com/tektoncd/chains/pkg/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+const (
+	StorageBackendCA  = "containeranalysis"
+	ProjectNameFormat = "projects/%s"
+	NoteNameFormat    = "projects/%s/notes/%s"
+)
+
+// Backend is a storage backend that stores signed payloads in the TaskRun metadata as an annotation.
+// It is stored as base64 encoded JSON.
+type Backend struct {
+	logger         *zap.SugaredLogger
+	tr             *v1beta1.TaskRun
+	client         pb.GrafeasV1Beta1Client
+	cfg            config.Config
+	occurrenceRefs []string // store occurrence IDs that are automatically generated during the time of creation
+}
+
+func NewStorageBackend(logger *zap.SugaredLogger, tr *v1beta1.TaskRun, cfg config.Config) (*Backend, error) {
+	// add TypeMeta information to taskrun objects
+	addTypeInformationToObject(tr)
```

Comment:
This is needed when we compose the URI for taskrun artifact to solve the issue that GroupVersionKind returns empty string. I thought it would be something general to add for a taskrun, and otherwise if we move it closer to URI generation, it would be called each time we store attestation of an artifact from the taskrun.

Reply:
My concern about doing this is if it's modifying the TaskRun, this could modify the checksum being produced depending on the ordering of when/how the object is signed.

The safest thing to do would be to not modify the incoming TaskRun, and only extract the values where needed.

Reply:
Great point! Yes, storage backends are created before the objects get signed in the `SignTaskRun` function ([source](https://github.com/tektoncd/chains/blob/ae2c66b7814e565c7a100f3849b9ec15337f0759/pkg/chains/signing.go#L127-L172)).
I'll move this to where it gets used (inside `StorePayload` where objects have been signed already).